File: Data Structures & Algorithms/target-sum/submission-11.java
class Solution {
    Integer[][] dp;
    int total;
    public int findTargetSumWays(int[] nums, int target) {
        total = Arrays.stream(nums).sum();
        dp = new Integer[nums.length][((total+1)*2)];
        return dfs(0,0,nums,target);

        
    }

    public int dfs(int i, int s,int[] nums,int target){
        if(i==nums.length){
            if(s==target){
                return 1;
            }
            return 0;
        }
        if(dp[i][s+total] != null){
            return dp[i][s+total];
        }
        dp[i][s+total] = dfs(i+1,s+nums[i],nums,target) + dfs(i+1,s-nums[i],nums,target);
        return dp[i][s+total];
    }
}

File: Data Structures & Algorithms/word-break/submission-11.py
class Solution:
    def wordBreak(self, s: str, wordDict: List[str]) -> bool:
        N = len(s)
        dp = set()
        def dfs(i):
            if i == N:
                return True
            if i in dp:
                return False
            for w in wordDict:
                if i+len(w)<=N and s[i:i+len(w)] == w and dfs(i+len(w)):
                    return True
            dp.add(i)
            return False
        return dfs(0)

        